core/src/test/java/org/elasticsearch/index/query/GeoBoundingBoxQueryBuilderTest.java

```diff
@@ -0,0 +1,374 @@
+        builder.coerce(true);
+        assertNull("Validation should not detect flipped topLeft/bottomRight lon corners: " + "topLeft: " + builder.topLeft() + " vs. bottomRight:"
+                + builder.bottomRight() + " ", builder.validate(false));
+        assertNull("Inner validation should not check for flipped lat corners.", builder.validate(true));
+        assertNull("Inner validation should not check for flipped lat corners.", builder.validate(false));
+
+        builder.coerce(false).ignoreMalformed(false);
+        assertNull("Validation should not detect flipped topLeft/bottomRight lon corners: " + "topLeft: " + builder.topLeft() + " vs. bottomRight:"
+                + builder.bottomRight() + " ", builder.validate(false));
+        assertNull("Inner validation should not check for flipped lat corners.", builder.validate(true));
+        assertNull("Inner validation should not check for flipped lat corners.", builder.validate(false));
+    }
+
+    @Test
+    public void testNormalization() throws IOException {
+        GeoBoundingBoxQueryBuilder qb = createTestQueryBuilder();
+        if (getCurrentTypes().length != 0 && "mapped_geo".equals(qb.fieldName())) {
+            // only execute this test if we are running on a valid geo field
+            qb.topLeft(200, 200);
+            qb.coerce(true);
+            Query query = qb.toQuery(createShardContext());
+            if (query instanceof ConstantScoreQuery) {
+                ConstantScoreQuery result = (ConstantScoreQuery) query;
+                BooleanQuery bboxFilter = (BooleanQuery) result.getQuery();
+                for (BooleanClause clause : bboxFilter.clauses()) {
+                    NumericRangeQuery boundary = (NumericRangeQuery) clause.getQuery();
+                    if (boundary.getMax() != null) {
+                        assertTrue("If defined, non of the maximum range values should be larger than 180", boundary.getMax().intValue() <= 180);
+                    }
+                }
+            } else {
+                assertTrue("memory queries should result in InMemoryGeoBoundingBoxQuery", query instanceof InMemoryGeoBoundingBoxQuery);
+            }
+        }
+    }
+
+    @Override
+    protected void doAssertLuceneQuery(GeoBoundingBoxQueryBuilder queryBuilder, Query query, QueryShardContext context) throws IOException {
+        if (getCurrentTypes().length == 0 || "mapped_geo".equals(queryBuilder.fieldName()) == false) {
+            assertTrue("For non geo fields expect MatchNoDocsQuery.", query instanceof MatchNoDocsQuery);
+        } else if (queryBuilder.type() == Type.INDEXED) {
+            assertTrue("Found no indexed geo query.", query instanceof ConstantScoreQuery);
+        } else {
+            assertTrue("Found no indexed geo query.", query instanceof InMemoryGeoBoundingBoxQuery);
+        }
+    }
+
+    // Java really could do with function pointers...
+    private interface PointTester {
```

Comment:
nitpick: this could be an abstrat class instead and have a coordinate instance member rather than repeating it all over.